integrated_channels/integrated_channel/learner_data.py

```diff
@@ -0,0 +1,215 @@
+        # Order by the course_id, to avoid fetching course API data more than we have to.
+        enrollment_queryset = EnterpriseCourseEnrollment.objects.select_related(
+            'enterprise_customer_user'
+        ).filter(
+            enterprise_customer_user__enterprise_customer=self.enterprise_customer,
+        ).order_by('course_id')
+
+        # Fetch course details from the Course API, and cache between calls.
+        course_details = None
+        course_api = None
+
+        for enterprise_enrollment in enrollment_queryset:
+
+            # Omit any enrollments where consent has not been granted
+            if not enterprise_enrollment.consent_available():
+                continue
+
+            course_id = enterprise_enrollment.course_id
+
+            # Fetch course details from Courses API
+            if course_details is None or course_details['course_id'] != course_id:
+                if course_api is None:
+                    course_api = CourseApiClient()
+                course_details = course_api.get_course_details(course_id)
+
+            if course_details is None:
+                # Course not found, so we have nothing to report.
+                LOGGER.error("No course details found for enrollment %d: %s",
+                             enterprise_enrollment.pk, course_id)
+                continue
+
+            # For instructor-paced courses, let the certificate determine course completion
+            if course_details.get('pacing') == 'instructor':
+                completed_date, grade = self._collect_certificate_data(enterprise_enrollment)
+
+            # For self-paced courses, check the Grades API
+            else:
+                completed_date, grade = self._collect_grades_data(enterprise_enrollment, course_details)
+
+            yield self.get_learner_data_record(
+                enterprise_enrollment=enterprise_enrollment,
+                completed_date=completed_date,
+                grade=grade,
+            )
+
+    def _collect_certificate_data(self, enterprise_enrollment):
+        """
+        Collect the learner completion data from a certificate.
+
+        Used for Instructor-paced courses.
+
+        If no certificate is found, then returns the completed_date = None, grade = In Progress, on the idea that a
+        certificate will eventually be generated.
+        """
+
+        if CourseKey is None or GeneratedCertificate is None:
+            raise NotConnectedToOpenEdX(_('This package must be installed in an OpenEdX environment.'))
+
+        try:
+            # TODO - use Certificates API instead?
+            certificate = GeneratedCertificate.eligible_certificates.get(
+                user__id=enterprise_enrollment.enterprise_customer_user.user_id,
+                course_id=CourseKey.from_string(enterprise_enrollment.course_id),
+            )
+
+            completed_date = certificate.created_date
+            # TODO: change to Pass/Fail/In Progress?
```

Comment:
@mattdrayer  Just need someone to tell me what the grade should be :)  I've got all the information I need from the certificate, so can show pass/fail instead of the certificate grade.